### DISCUSSION TITLE
extending-the-shutdown-timeout-setting-to-ensure-graceful-ihostedservice-shutdown/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#100] [Announcements]
# extending-the-shutdown-timeout-setting-to-ensure-graceful-ihostedservice-shutdown/

In this post I describe a problem where you get an exception when shutting down an app that used IHostedService, when the service takes too long to shut down.

https://andrewlock.net/extending-the-shutdown-timeout-setting-to-ensure-graceful-ihostedservice-shutdown/

**Comment by grewegreg** [comment]
Today I stumbled across this article and found it very helpful.   Though I do have a question, instead of creating a new Cancellation Token in the StopAsync method, why not create a Cancellation Token at startup, and pass it to the RunAysnc method?

Something like:
~~~
HostOptions hostOptions= AppConfigHelper.GetHostOptions();
using CancellationTokenSource cts = new(hostOptions.ShutdownTimeout);
await IHostConfiguration.CreateHostBuilder(args).Build().RunAsync(cts.token);
~~~

**Reply by grewegreg** [reply]
Yesterday I did try to implement the pseudo code listed above and the application shutdown after the timeout value was exceeded.  After some research, this would be the expected behavior.  It seems that each task will need to have its own cancellation token, which makes total sense.

[a comment by an automated account is not shown]

**Comment by asimmon** [comment]
It might be worth mentioning that since .NET 7 the default `ShutdownTimeout` has been increased from 5 seconds to 30 seconds.

https://github.com/dotnet/runtime/blob/v6.0.0/src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs#L18
https://github.com/dotnet/runtime/blob/v7.0.0/src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs#L18

**Comment by terryaney** [comment]
Is it possible to prevent shutdown via `BackgroundService.StopAsync()`? Or maybe a better pattern exists, but let's assume I've triggered 4-5 long running processes from my service. There is really no indication in the Services MMC snap-in, so people request a stop, but I want to prevent the stop from happening if there are jobs running. I've implemented a sliding window so that it'll forcibly shutdown if enough repeated requests are sent, but it is just a safety precaution. I had this implemented in .NET Framework/Topshelf  via `Topshelf.ServiceControl.Stop()` and trying to replicate in .NET 7 upgrade.

**Comment by carlin-q-scott** [comment]
Did you know that the SDK [binds the `HostOptions` to the root configuration](https://github.com/dotnet/dotnet/blob/main/src/runtime/src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs#L328) using [slightly different setting names](https://github.com/dotnet/dotnet/blob/main/src/runtime/src/libraries/Microsoft.Extensions.Hosting/src/HostOptions.cs#L64):

```yaml
startupTimeoutSeconds:
shutdownTimeoutSeconds:
servicesStartConcurrently:
servicesStopConcurrently:
```
